### ISSUE TITLE
[apex] UnusedMethod does not report any unused methods (7.0.0-rc4)

### ISSUE THREAD

**Issue by nwcm** [pmd/pmd#4810]
**Affects PMD Version:** 7.0.0-rc4

**Rule:** [UnusedMethod](https://docs.pmd-code.org/latest/pmd_rules_apex_design.html#unusedmethod)

**Description:**
The Apex Unused Method rule does not report any private or public methods that are unused. With or without `PMD_APEX_ROOT_DIRECTORY` set.

**Code Sample demonstrating the issue:**

File to run rules on
```java
public with sharing class testUnusedMethod {
    public testUnusedMethod() {

    }

    public static void unusedPublic(){
        return;
    }

    private static void unusedPrivate(){
        return;
    }

    public static void usedPublic(){
        return;
    }

    private static void usedPrivate(){
        return;
    }


    private static void use(){
        usedPrivate();
    }

}
```

Class referencing public method
```java
public with sharing class testUnusedMethodUse {
    public testUnusedMethodUse() {

    }

    private static void useTheMethod(){
        testUnusedMethod.usedPublic();
    }
}
```

**Expected outcome:**

`unusedPrivate` should always be thrown as unusedMethod as it is not referenced in the class

`unusedPublic` should be thrown as unusedMethod if `PMD_APEX_ROOT_DIRECTORY` is correctly set as it is not referenced from this or any other classes

PMD should report a violation at line ..., but doesn't. This is a false-negative.

**Running PMD through:** *CLI*

**Additional**  
In the my source when `PMD_APEX_ROOT_DIRECTORY` is set. We see many INFO logs from ApexMultifileAnalysis. 

`[main] INFO net.sourceforge.pmd.lang.apex.multifile.ApexMultifileAnalysis - C:\Users\foo\repos\bar\sfdx-project.json: Error: line 1: Extending unknown SObject error:`

We extend other managed packages with additional fields from our source, as such ApexMulifileAnalysis seems to fall over trying to work with the fields that don't have a related object definition

```
pmd --version
PMD 7.0.0-rc4

java --version
openjdk 17.0.9 2023-10-17
OpenJDK Runtime Environment Temurin-17.0.9+9 (build 17.0.9+9)
OpenJDK 64-Bit Server VM Temurin-17.0.9+9 (build 17.0.9+9, mixed mode, sharing)
```

**Comment by jsotuyod** [comment]
Hi @nwcm would you mind submitting the contents of your `sfdx-project.json` file?

Similarly, a full log when running with `-v` (debug mode) would be useful to better understand what may be going on.

**Comment by jsotuyod** [comment]
@nwcm thanks for that.

This is obviously redacted, but the references to `ThirdPartyNamespace` makes me wonder…

As per [the docs](https://docs.pmd-code.org/latest/pmd_rules_apex_design.html#unusedmethod):

> If the Apex code references external packages via namespace(s) you should declare these in your sfdx-project.json file using the ‘plugins’ syntax shown in the example below to avoid errors. Here’s an example of a well-formed sfdx-project.json:
```javascript
{
    "packageDirectories": [
      {
        "path": "src",
        "default": true
      }
    ],
    "namespace": "my_namespace",
    "sfdcLoginUrl": "https://login.salesforce.com",
    "sourceApiVersion": "52.0",
    "plugins": {
        "dependencies": [
            {"namespace": "aa"}
        ]
    }
}
```

Maybe this is causing your issues?

**Comment by jsotuyod** [comment]
@nwcm awesome! Happy to hear that!
